### ISSUE TITLE
How do I set a trigger?

### ISSUE THREAD

**Issue by hdroller** [derrian-distro/LoRA_Easy_Training_Scripts#149]
I can't figure out how to set the trigger. I've tried adding the trigger word to the beginning list of the tags in each txt file but it didn't work. Any suggestions?

**Comment by derrian-distro** [comment]
Just make sure to set keep tokens to however many you have at the front of the txt files, that will keep it at the front, making it an activation tag

**Comment by derrian-distro** [comment]
no problem, to be fair, lora don't actually "activate" they just apply and are always active. that's why the trigger word is actually just teaching the model that this keyword holds importance to the weights. it's pretty likely that you just haven't trained it enough if you aren't getting a result as expected

**Comment by Jelosus2** [comment]
Could i ask 

> Ok, so I retried my training but with "keep tokens" set to 1 since the first token is what I want for the activation tag. But it's still not activating. I'm going to try tweaking my LR and see if that makes a difference. Thanks for the quick response btw.

Could I know what settings you used for the training process and the number of images you have for the training?

**Comment by hdroller** [comment]
Here's my yoml file. My LR is probably too high but it does the job. By default it's set to 0.0001 so I cranked it up to 0.005

[[subsets]]
num_repeats = 3
keep_tokens = 1
caption_extension = ".txt"
shuffle_caption = true
flip_aug = false
color_aug = false
random_crop = false
is_reg = false
image_dir = "D:/iruma/2/ameri"

[noise_args]

[sample_args]

[logging_args]

[general_args.args]
pretrained_model_name_or_path = "E:/EZ Lora/LoRA_Easy_Training_Scripts/NAI-final-pruned.ckpt"
mixed_precision = "fp16"
seed = 23
clip_skip = 2
max_data_loader_n_workers = 1
persistent_data_loader_workers = true
max_token_length = 225
prior_loss_weight = 1.0
xformers = true
max_train_epochs = 4
vae = ""

[general_args.dataset_args]
resolution = [ 768, 512,]
batch_size = 1

[network_args.args]
network_dim = 128
network_alpha = 64.0
min_timestep = 0
max_timestep = 1000

[optimizer_args.args]
optimizer_type = "AdamW8bit"
lr_scheduler = "cosine_with_restarts"
learning_rate = 0.005
max_grad_norm = 1.0
lr_scheduler_num_cycles = 1

[saving_args.args]
output_dir = "E:/EZ Lora"
save_precision = "fp16"
save_model_as = "safetensors"
output_name = "ameri"

[bucket_args.dataset_args]
enable_bucket = true
min_bucket_reso = 320
max_bucket_reso = 1024
bucket_reso_steps = 64

[optimizer_args.args.optimizer_args]
weight_decay = "0.1"
betas = "0.9,0.99"
decouple = "true"

**Comment by derrian-distro** [comment]
Ok, so I see a few things of note, namely, you use cosine with restarts but don't restart at all, and your dim is significantly higher than you need it to be, I use dim 16 for my characters for example, and those usually include more than one outfit.

Your lr *is* too high, but because you don't restart, it reaches really low quick enough to not matter. I usually set my unet to 5e-4 (0.0005) and my te to 1e-4 (0.0001)

Why train at 768×512? Wouldn't it be better to train at 768x768? Or is this because of a vram limit?

Usually, a higher batch size than 1 is better. At least 2 helps a lot in preventing baking.

Also, decouple is for dadapt only. It doesn't do anything on adamw8bit

**Comment by hdroller** [comment]
Yeah I tried to follow some guides but got mixed up on a few things. Thanks for the tips, i'm new to this.
- yeah i forgot to add restarts
- i'll try a low dim and LR instead
- I trained 768x512 because that was the size of my screenshots. 
- batch/epic/resets confuse me so i left it at 1
- and i forgot to remove decouple as i tried a different optimizer before.

**Comment by derrian-distro** [comment]
If all of your images are in that resolution, then it's fine to keep it that way

I suggest either 2 or 3 on the restart number, with a lower unet lr of 5e-4. As it stands, it's likely that most of the steps you trained were actually at lr 0.

The te generally doesn't need to be trained nearly as heavily as the unet, so it might even be worthwhile to have the te even lower than 1e-4, 5e-5 is generally accepted.

The way you can calculate steps is following this algorithm (num images * repeats) / batch size)

And very fair on decouple, it doesn't really affect much so it was just something to point out as to not mislead you

On average, dim 16 gets nearly every character I've made 100% except for when I tried to fit 6 or 7 outfits into one lora, in which case made it a locon with a low dim 4 conv dim.

If you have any other questions I'll try to answer them when I can

**Comment by hdroller** [comment]
Will do, thanks for the info!

**Comment by hdroller** [comment]
I forgot to ask, is the repeats in the algorithm you mentioned referring to the repeats in the optimizer or in the subset? I see there's two different fields for resets "Num Restarts" is the optimizer reset and "Number of Restarts" in the Subset Args tab.

**Comment by derrian-distro** [comment]
the subsets, num restarts is for the cosine with restarts

**Comment by Jelosus2** [comment]
> Here's my yoml file. My LR is probably too high but it does the job. By default it's set to 0.0001 so I cranked it up to 0.005. I have about 300 images for my character.
> 
> [[subsets]] num_repeats = 3 keep_tokens = 1 caption_extension = ".txt" shuffle_caption = true flip_aug = false color_aug = false random_crop = false is_reg = false image_dir = "D:/iruma/2/ameri"
> 
> [noise_args]
> 
> [sample_args]
> 
> [logging_args]
> 
> [general_args.args] pretrained_model_name_or_path = "E:/EZ Lora/LoRA_Easy_Training_Scripts/NAI-final-pruned.ckpt" mixed_precision = "fp16" seed = 23 clip_skip = 2 max_data_loader_n_workers = 1 persistent_data_loader_workers = true max_token_length = 225 prior_loss_weight = 1.0 xformers = true max_train_epochs = 4 vae = ""
> 
> [general_args.dataset_args] resolution = [ 768, 512,] batch_size = 1
> 
> [network_args.args] network_dim = 128 network_alpha = 64.0 min_timestep = 0 max_timestep = 1000
> 
> [optimizer_args.args] optimizer_type = "AdamW8bit" lr_scheduler = "cosine_with_restarts" learning_rate = 0.005 max_grad_norm = 1.0 lr_scheduler_num_cycles = 1
> 
> [saving_args.args] output_dir = "E:/EZ Lora" save_precision = "fp16" save_model_as = "safetensors" output_name = "ameri"
> 
> [bucket_args.dataset_args] enable_bucket = true min_bucket_reso = 320 max_bucket_reso = 1024 bucket_reso_steps = 64
> 
> [optimizer_args.args.optimizer_args] weight_decay = "0.1" betas = "0.9,0.99" decouple = "true"

may i know how many images do you have for the concept?

**Comment by hdroller** [comment]
about 110

**Comment by Jelosus2** [comment]
> about 110

try 2 repeats, 512x512 or 768x768, learning rate 1e-4, unet 5e-4, text encoder 1e-4, dim/alpha depends but usually 32/16 or 16/8, warmup ratio: 0.05 (5%), cycles for scheduler: 3 and batch size 2 if your PC allows it and 10 epochs
